### DISCUSSION TITLE
Divider for External Tools ribbon

### DISCUSSION THREAD

**Discussion by JamesDBartlett3** [didierterrien/Power-BI-Sidetools#16] [Show and tell]
Hi Didier,

I have created an External Tool called **Divider**, which is just a slim vertical bar that divides or clusters the External Tools ribbon into groups or categories, like this:

![image](https://user-images.githubusercontent.com/37491308/118333840-1138d080-b4d2-11eb-9d37-f96456d54543.png)
![image](https://user-images.githubusercontent.com/37491308/118333802-fd8d6a00-b4d1-11eb-9f26-9ce548112204.png)

Would you be interested in adding an option in your **External Tool Wizard** to insert one or more instances of **Divider** into the External Tools ribbon?

![image](https://user-images.githubusercontent.com/37491308/118334484-53164680-b4d3-11eb-92a1-211ad1ac90cc.png)

Cheers,

James

**Comment by didierterrien** [comment]
Hi James,
That's a great idea !  Sure I will add it to the wizard. 
Is it possible to make the image as narrow as possible in order to save the place for icons?
If you want me to add some credits, feel free to send a short text.

**Reply by JamesDBartlett3** [reply]
The icon is a transparent PNG at 300x300px. I tried making it narrower, but it wouldn't display properly. I think there must be a software limitation in the part of Power BI Desktop that renders the icons.

**Comment by didierterrien** [comment]
Do you have an idea what could be the action if somebody clicks on the icon?

**Reply by JamesDBartlett3** [reply]
If a user clicks the icon, it will not do anything visible to the user. The **Path** field can't be left blank, and it throws an error if you give it a path that doesn't resolve, so I put rundll32.exe in there with no arguments, so it'll run for a fraction of a second, but nothing appears on the screen, and no errors are shown. If there's a better way to do this, I'm totally open to ideas.

**Comment by JamesDBartlett3** [comment]
Here's a zip file containing the source code for Divider:
[Divider.zip](https://github.com/didierterrien/Power-BI-Sidetools/files/6487698/Divider.zip)

One other thing: The screenshot above shows the word "Divider" in the Tool Name field, but it's necessary to blank that out before saving it if you don't want the word "Divider" to show up as a label under the bar. I used a non-printing character, because it wouldn't accept whitespace. You won't see the character in the json file, but if you copy and paste it into PowerShell and check its length, you'll see that it's definitely there. 

![image](https://user-images.githubusercontent.com/37491308/118370278-996bb400-b56c-11eb-8a1d-36daf9bc9e78.png)

**Reply by didierterrien** [reply]
Thanks James,
I will add it to the wizard mentionning your name in the description except if you don't want.

**Reply by JamesDBartlett3** [reply]
I'm happy to contribute, even if it's something small like this, and I hope to learn more about programming for Power BI External Tools, so I can contribute even more in the future. Sure, you can add my name in the credits if you want. ☺️

**Comment by didierterrien** [comment]
Hello James,
The feature will appear in version 1.2.5
With a binary editor, I guessed the invisible string is this one : &HE2    &H80    &H8E
I tried it but it displays weird characters. So at the moment I replaced it with a dot.
I removed the path to rundll32.exe so it will not fail with different installations
Thanks a lot for your contribution. That's a great idea !

**Reply by didierterrien** [reply]
I even like more the \ character. Need to double it in the Json ("name": "\\\\")
![image](https://user-images.githubusercontent.com/37301981/129388744-456a54b2-7eaf-481f-bef0-5f1e6d43722a.png)

**Reply by JamesDBartlett3** [reply]
@didierterrien Is something wrong with the non-printing character? I tried lots of different visible characters when I was making it, but eventually settled on the non-printing character because it consumes the least possible space.

**Reply by JamesDBartlett3** [reply]
By the way, the non-printing character is [Unicode U+200E Left-to-Right Mark](https://www.fileformat.info/info/unicode/char/200e/index.htm). I just copied and pasted it into [this](https://babelstone.co.uk/Unicode/whatisit.html) to identify it.

**Comment by didierterrien** [comment]
Hello James,
I found the issue with the non-printing character. So from version 1.2.5, it will be possible to hide any external tool name.
I love this new feature. Thanks again

**Reply by JamesDBartlett3** [reply]
Yay! I'm excited for the new version! Thanks for letting me help to add a useful feature to Power BI Sidetools!

**Comment by didierterrien** [comment]
Power BI Sidetools has been updated to version 1.2.5

**Reply by JamesDBartlett3** [reply]
Hi Didier,

I just installed version 1.2.5, and I noticed a few things:

1. When I add a Separator from the Power BI Sidetools External Tool Manager, and click on the Separator icon, it throws an error: ![image](https://user-images.githubusercontent.com/37491308/131570568-0dd989ff-6883-446b-8818-e2d99ddad416.png) However, when I click on the icon for my original `divider.pbitool.json`, there is no error. I believe the reason for the error is because the full path to `rundll32.exe` was removed. I discovered while developing the divider that I had to include the full path to `C:\windows\system32\rundll32.exe` to avoid throwing an error if the user clicks on the icon.
2. It looks like the vertical bar in the Separator that was added by Power BI Sidetools is thicker than in the one I made. Was that intentional?
3. I found a small typo in the Description text:
     > Vertical section divider bar from James Bartlett. ~~Usefull~~ Useful for clustering external tools into groups or categories.

Please let me know if there's anything I can do to help!

James

**Reply by didierterrien** [reply]
Hi James,
This path didn't work on my computer (C:\windows\system32\rundll32.exe) because the file is located in an other folder so I removed the directories and then it worked. I thought it was more universal but it probably depends on environment variables. We need to find an universal solution. I can easily open external tools manager when the user clicks on the separator or we can run a CMD command.

I prefer to include a png file in the project instead of a base64 string so I tried to create the same image than yours. Feel free to send a png image. I can easily change it.

The typo is corrected. Thank you !

Thanks for your collaboration !

**Reply by JamesDBartlett3** [reply]
Hi Didier,

> This path didn't work on my computer (C:\windows\system32\rundll32.exe) because the file is located in an other folder so I removed the directories and then it worked. I thought it was more universal but it probably depends on environment variables. We need to find an universal solution. I can easily open external tools manager when the user clicks on the separator or we can run a CMD command.
  
Can you help me understand why `rundll32.exe` is not in your `C:\Windows\system32` folder? That's where it exists on every Windows system I've ever worked with, so I don't understand why yours doesn't find it. 
  
I tried installing Power BI Sidetools on a separate machine to see if maybe the `%path%` environment variable on my primary machine got messed up, but I have the same exact problem on the second machine. I also tried replacing `C:\Windows` in the path with the `%SystemRoot%` environment variable, and that didn't work either. It seems to me that on both of the machines where I tried it, the External Tools in Power BI Desktop cannot access the host system's environment variables, and that's why they cannot find `rundll32.exe` without the full path to it.
  
>I prefer to include a png file in the project instead of a base64 string so I tried to create the same image than yours.
  
When I open the `---------------.pbitool.json` file that was created by Power BI Sidetools, it has a base64 string for the icon, not a png file, so I'm not sure what's going on there. Here's a screenshot showing my file next to yours:
  
![image](https://user-images.githubusercontent.com/37491308/132076883-aefc563c-56e5-41c4-8652-822896404a14.png)
  
>Feel free to send a png image. I can easily change it.
  
I actually included the original image that I used to create the base64-encoded string in the original [Divider.zip](https://github.com/didierterrien/Power-BI-Sidetools/files/6487698/Divider.zip) file that I attached earlier in this thread. The file is called Divider.png, and when I created the `009-divider.pbitool.json` file, I uploaded that image at [this website](https://www.base64-image.de/ ) to convert it to base64 encoding, and copy-pasted the result into the file. Just now, I tried uploading the same image to [two](https://elmah.io/tools/base64-image-encoder/) [other](https://base64.guru/converter/encode/image) base64 encoding websites, and I got the exact same string from all of them, so I'm not sure how it ended up looking so different between mine and yours, but the difference is pretty stark.
  
![image](https://user-images.githubusercontent.com/37491308/132078481-419d4e2c-54f6-4d07-be16-9c33d0afe852.png)
  
That's yours on the left, and mine on the right.

Thanks for taking the time to troubleshoot this with me, and I hope you have a pleasant weekend!
  
James

**Reply by didierterrien** [reply]
Hello James,
In version 1.2.6 :
- I modified the path to "C:\windows\System32\rundll32.exe"
- I switched the tool name visibility to false by default
- I updated the png file with the one in your Divider.zip
- I corrected the typo

This version will be available for tests soon
Have a good Sunday ! Mine is very sunny
Didier

**Reply by JamesDBartlett3** [reply]
That's great news, Didier! Thanks for the update. 
Let me know if you could use any assistance with testing version 1.2.6, and I'd be happy to help.

**Comment by didierterrien** [comment]
Hello James, I will be very happy if you can help with tests. New exciting features are coming and your opinion will be precious.

I published V1.2.6 test version. Can you try it ? Feel free to post new issues.

Test version 1.2.6 is meant to 
- evaluate the new auto update feature. It tries systematically to update to the production version. When you will finish to evaluate V1.2.6, please try to update and provide feedback about possible issues.
- improve connection with reports connected to Power BI datasets
- evaluate the preview version of DAX generator

DAX generator provides 2 files repositories containing DAX formulas. One contains your own DAX templates (user templates) and the second one contains DAX templates written by experts (featured templates).
Those templates contain	a DAX formula and several instructions for the DAX generator. Instructions declare parameters so the user can change the measure name, tables name, columns names, values in few clicks.
Then the user generates the DAX formula which can be pasted in Power BI desktop, Tabular Editor, DAX Studio, ... DAX generator provides the possibility to create or update measures with one click in the desktop. An other click allows to debug DAX formulas in DAX debugger.

**Reply by JamesDBartlett3** [reply]
Thanks, Didier! I'm downloading it now, and I'll definitely post an issue if I find any bugs. Cheers, and have a great weekend!
